src-electron/rest/user-data.js

```diff
@@ -84,12 +84,123 @@ function httpGetEndpointIds(db) {
+    let { deviceTypeRefs, endpointTypeRef } = request.query
+    if (Array.isArray(deviceTypeRefs) && deviceTypeRefs.length > 0) {
+      let deviceTypeFeatures = await queryFeature.getFeaturesByDeviceTypeRefs(
+        db,
+        deviceTypeRefs,
+        endpointTypeRef
+      )
+      response.status(StatusCodes.OK).json(deviceTypeFeatures)
+    } else {
+      response.status(StatusCodes.OK).json([])
+    }
+  }
+}
+
+/**
+ * Get all attributes, commands and events in an endpoint type cluster.
+ * @param {*} db
+ * @param {*} endpointTypeClusterId
+ * @param {*} deviceTypeClusterId
+ * @returns elements object containing all attributes, commands and events
+ * in an endpoint type cluster
+ */
+async function getEndpointTypeElements(
+  db,
+  endpointTypeClusterId,
+  deviceTypeClusterId
+) {
+  let elements = {}
+  elements.attributes =
+    await queryAttribute.selectAttributesByEndpointTypeClusterId(
+      db,
+      endpointTypeClusterId
+    )
+  elements.commands =
+    await queryCommand.selectCommandsByEndpointTypeClusterIdAndDeviceTypeClusterId(
+      db,
+      endpointTypeClusterId,
+      deviceTypeClusterId
+    )
+  elements.events =
+    await queryEvent.selectEventsByEndpointTypeClusterIdAndDeviceTypeClusterId(
+      db,
+      endpointTypeClusterId,
+      deviceTypeClusterId
+    )
+  return elements
+}
+
+/**
+ * HTTP POST: elements to be updated after toggle a device type feature
+ *
+ * @param {*} db
+ * @returns callback for the express uri registration
+ */
+function httpPostCheckConformOnFeatureUpdate(db) {
+  return async (request, response) => {
+    let sessionId = request.zapSessionId
+    let { featureData, featureMap, endpointId } = request.body
+    let { endpointTypeClusterId, deviceTypeClusterId } = featureData
+
+    let elements = await getEndpointTypeElements(
+      db,
+      endpointTypeClusterId,
+      deviceTypeClusterId
+    )
+    // check element conform and return elements that need to be updated
+    let result = queryFeature.checkElementConformance(
+      elements,
+      featureMap,
+      featureData,
+      endpointId
+    )
+
+    // set device type feature warning
+    await querySessionNotification.setNotificationOnFeatureChange(
+      db,
+      sessionId,
+      result
+    )
+    // do not set element warning if feature change disabled
+    if (!result.disableChange) {
+      let outdatedWarnings = queryFeature.getOutdatedElementWarning(
+        featureData,
+        elements,
+        result.elementMap
+      )
+      await querySessionNotification.deleteNotificationWithPatterns(
+        db,
+        sessionId,
+        outdatedWarnings
+      )
+    }
+
+    response.status(StatusCodes.OK).json(result)
+  }
+}
+
+/**
+ * HTTP POST: required and unsupported cluster elements based on conformance
+ * @param {*} db
+ * @returns callback for the express uri registration
+ */
+function httpPostSetRequiredElements(db) {
```

Comment:
I'm confused with the name of this function..... Why do you need to POST the required elements? Aren't the required elements calculated on the back-end and at best sent over to the front-end as result of some "POST" that sets feature bits or something like that?

Reply:
You are right. I changed it to a GET request as it only gets resources from backend and not change them